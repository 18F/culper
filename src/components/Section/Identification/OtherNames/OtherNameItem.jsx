import React from 'react'
import { i18n } from '../../../../config'
import { Field, MaidenName, Name, Textarea, DateRange } from '../../../Form'

export default class OtherNameItem extends React.Component {
  constructor (props) {
    super(props)
    this.update = this.update.bind(this)
    this.updateName = this.updateName.bind(this)
    this.updateMaidenName = this.updateMaidenName.bind(this)
    this.updateDatesUsed = this.updateDatesUsed.bind(this)
    this.updateReason = this.updateReason.bind(this)
  }

  update (queue) {
    this.props.onUpdate({
      Name: this.props.Name,
      MaidenName: this.props.MaidenName,
      DatesUsed: this.props.DatesUsed,
      Reason: this.props.Reason,
      ...queue
    })
  }

  updateName (values) {
    this.update({
      Name: values
    })
  }

  updateMaidenName (values) {
    this.update({
      MaidenName: values
    })
  }

  updateDatesUsed (values) {
    this.update({
      DatesUsed: values
    })
  }

  updateReason (values) {
    this.update({
      Reason: values
    })
  }

  render () {
    return (
      <div className="other-name">
        <Field title={i18n.t('identification.othernames.heading.name')}
               optional={true}
               scrollIntoView={this.props.scrollIntoView}>
          <Name name="Name"
                {...this.props.Name}
                onUpdate={this.updateName}
                required={this.props.required}
                scrollIntoView={this.props.scrollIntoView}
                />
        </Field>

        <Field title={i18n.t('identification.othernames.heading.maiden')}
               help="alias.maiden.help"
               adjustFor="buttons"
               scrollIntoView={this.props.scrollIntoView}
               shrink={true}>
          <MaidenName name="MaidenName"
                      required={this.props.required}
                      {...this.props.MaidenName}
                      onUpdate={this.updateMaidenName}
                      />
        </Field>

        <Field title={i18n.t('identification.othernames.heading.used')}
               help="alias.used.help"
               adjustFor="daterange"
               scrollIntoView={this.props.scrollIntoView}
               shrink={true}>
          <DateRange name="DatesUsed"
                     {...this.props.DatesUsed}
                     applicantBirthdate={this.props.applicantBirthdate}
                     onUpdate={this.updateDatesUsed}
                     required={this.props.required}
                     />
        </Field>

        <Field title={i18n.t('identification.othernames.heading.reason')}
               scrollIntoView={this.props.scrollIntoView}
               help="alias.reason.help">
          <Textarea name="Reason"
                    {...this.props.Reason}
                    onUpdate={this.updateReason}
                    className="reason"
                    required={this.props.required}
                    />
        </Field>
      </div>
    )
  }
}

OtherNameItem.defaultProps = {
  applicantBirthdate: {},
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr },
  required: false
}
