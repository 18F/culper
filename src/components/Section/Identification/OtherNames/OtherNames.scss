.other-names {
    h2 {
        margin-top: 2rem;
    }

    .maiden-name,
    .daterange,
    .reason {
        display: inline-block;
    }

    .reason {
        width: 90%;
        max-width: 64rem !important;

        input {
            width: 100% !important;
            max-width: 64rem !important;
        }
    }

    .maiden-help-icon {
        margin-left: -1.6rem;
        margin-top: 1rem;
    }
    
    .used-help-icon {
        vertical-align: top;
        margin-top: 20rem;
    }

    .reason-help-icon {
        vertical-align: top;
        margin-top: 10rem;
    }
}
