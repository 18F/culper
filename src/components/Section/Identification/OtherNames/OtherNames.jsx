import React from 'react'
import { i18n } from '../../../../config'
import schema from '../../../../schema'
import { Summary, NameSummary, DateSummary } from '../../../Summary'
import validate, { OtherNameValidator } from '../../../../validators'
import SubsectionElement from '../../SubsectionElement'
import { Field, Accordion, Textarea, DateRange, Branch, Show } from '../../../Form'
import OtherNameItem from './OtherNameItem'

export default class OtherNames extends SubsectionElement {
  constructor (props) {
    super(props)

    this.update = this.update.bind(this)
    this.updateBranch = this.updateBranch.bind(this)
    this.updateList = this.updateList.bind(this)
  }

  update (queue) {
    this.props.onUpdate({
      List: this.props.List,
      HasOtherNames: this.props.HasOtherNames,
      ...queue
    })
  }

  updateBranch (values) {
    this.update({
      HasOtherNames: values,
      List: values.value === 'Yes' ? this.props.List : { items: [] }
    })
  }

  updateList (values) {
    this.update({
      List: {
        items: values.items
      }
    })
  }

  /**
   * Assists in rendering the summary section.
   */
  summary (item, index) {
    const itemObj = item.Item || {}
    const dates = DateSummary(itemObj.DatesUsed)
    const name = NameSummary(itemObj.Name)

    return Summary({
      type: i18n.t('identification.othernames.collection.summary.name'),
      index: index,
      left: name,
      right: dates,
      placeholder: i18n.m('identification.othernames.collection.summary.unknown')
    })
  }

  render () {
    return (
      <div className="other-names">
        <Field title={i18n.t('identification.othernames.title')}
               titleSize="h2"
               optional={true}
               help="identification.othernames.branch.help"
               className="no-margin-bottom">
          {i18n.m('identification.othernames.info')}
        </Field>

        <Branch name="has_othernames"
                label={i18n.t('identification.othernames.branch.question')}
                labelSize="h3"
                {...this.props.HasOtherNames}
                warning={true}
                onUpdate={this.updateBranch}
                required={this.props.required}
                scrollIntoView={this.props.scrollIntoView}
                onError={this.handleError}>
        </Branch>
        <Show when={this.props.HasOtherNames.value === 'Yes'}>
          <Accordion {...this.props.List}
                     defaultState={this.props.defaultState}
                     onUpdate={this.updateList}
                     onError={this.handleError}
                     required={this.props.required}
                     validator={OtherNameValidator}
                     summary={this.summary}
                     description={i18n.t('identification.othernames.collection.summary.title')}
                     appendLabel={i18n.t('identification.othernames.collection.append')}>
                     <OtherNameItem name="Item"
                       required={this.props.required}
                       scrollIntoView={this.props.scrollIntoView}
                       bind={true}
                     />
          </Accordion>
        </Show>
      </div>
    )
  }
}

OtherNames.defaultProps = {
  List: {
    items: []
  },
  HasOtherNames: {},
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr },
  section: 'identification',
  subsection: 'othernames',
  dispatch: () => {},
  validator: (state, props) => {
    return validate(schema('identification.othernames', props))
  },
  defaultState: true,
  required: false
}
