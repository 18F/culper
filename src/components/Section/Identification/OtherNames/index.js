import OtherNames from './OtherNames'
export default OtherNames
