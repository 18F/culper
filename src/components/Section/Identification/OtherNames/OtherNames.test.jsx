import React from 'react'
import { shallow, mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'

import { OtherNames } from './OtherNames'

describe('The OtherNames component', () => {
  const mockStore = configureMockStore()

  const mountComponentWithStore = (props = {}, defaultState = {}) => {
    const store = mockStore({ ...defaultState })
    return mount(
      <Provider store={store}>
        <OtherNames {...props} />
      </Provider>
    )
  }

  it('renders without errors', () => {
    const component = shallow(<OtherNames />)

    expect(component.exists()).toBe(true)
    expect(component).toMatchSnapshot()
  })

  it('implements an onUpdate handler', () => {
    let updates = 0
    const props = {
      onUpdate: () => {
        updates += 1
      },
    }

    const component = mountComponentWithStore(props)
    component.find('.branch .yes input').simulate('change')
    expect(updates).toEqual(1)
  })

  describe('default state', () => {
    it('has no names initially', () => {
      const component = shallow(<OtherNames />)
      expect(component.find('.first input').length).toEqual(0)
    })
  })

  describe('When Yes is selected', () => {
    const props = {
      HasOtherNames: { value: 'Yes' },
      List: {
        items: [{ Item: {} }],
      },
    }

    const component = mountComponentWithStore(props)

    it('renders the form', () => {
      expect(component.find('.details').length).toBeGreaterThan(0)
    })

    it('displays inputs', () => {
      expect(component.find('.first input').length).toEqual(1)
    })

    it.skip('adds a name when button is clicked then collapses', () => {
      // TODO - this should be implemented as a test of the Accordion component
      component.find('.addendum .branch .yes').simulate('click')
    })
  })

  describe('When No is selected', () => {
    const props = {
      HasOtherNames: { value: 'No' },
    }

    const component = mountComponentWithStore(props)

    it('does not render the form', () => {
      expect(component.find('.details').length).toBe(0)
    })

    it('does not display any fields', () => {
      expect(component.find('.first input').length).toEqual(0)
    })
  })
})
