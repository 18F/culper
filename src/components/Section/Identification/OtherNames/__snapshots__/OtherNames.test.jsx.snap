// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The OtherNames component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="section-content other-names"
    data-section="identification"
    data-subsection="othernames"
  >
    <h1
      className="section-header"
    >
      Other names used
    </h1>
    <Field
      adjustFor=""
      className="no-margin-bottom"
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.othernames.branch.help"
      helpActive={false}
      optional={true}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Provide your other names used and the period of time you used them"
      titleSize="h3"
      validate={true}
    >
      <ReactMarkdown
        containerTagName="div"
        parserOptions={Object {}}
        renderers={
          Object {
            "link": [Function],
          }
        }
        source="For example: your maiden name, name(s) by a former marriage, former name(s), alias(es), or nickname(s)."
      />
    </Field>
    <Branch
      adjustFor="buttons"
      className=""
      confirmation="Are you sure you would like to change your answer? This will remove any data currently entered."
      disabled={false}
      label="Have you used any other names?"
      labelSize="h4"
      name="has_othernames"
      noAriaLabel=""
      noLabel="No"
      noValue="No"
      onError={[Function]}
      onUpdate={[Function]}
      optional={false}
      required={false}
      scrollIntoView={false}
      value=""
      warning={true}
      yesAriaLabel=""
      yesLabel="Yes"
      yesValue="Yes"
    />
    <Show
      when={false}
    >
      <Accordion
        appendClass=""
        appendHelp={null}
        appendLabel="Add another name"
        appendMessage={null}
        appendTitle="Do you have additional names to enter?"
        branch={Object {}}
        byline={[Function]}
        caption={null}
        className=""
        closeLabel="Close"
        customDetails={[Function]}
        customSummary={[Function]}
        defaultState={true}
        description="Summary of other names used"
        incomplete="There is a problem with this item, please open it and fix the issue(s)."
        initial={true}
        inject={[Function]}
        items={Array []}
        minimum={1}
        onError={[Function]}
        onUpdate={[Function]}
        openLabel="Open"
        realtime={true}
        removeLabel="Remove"
        required={false}
        scrollToBottom=""
        scrollToTop=""
        skipWarning={false}
        sort={null}
        summary={[Function]}
        timeout={500}
        titleSize="h4"
        transformer={[Function]}
        validator={[Function]}
      >
        <OtherNameItem
          bind={true}
          name="Item"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
        />
      </Accordion>
    </Show>
  </div>,
  "nodes": Array [
    <div
      className="section-content other-names"
      data-section="identification"
      data-subsection="othernames"
    >
      <h1
        className="section-header"
      >
        Other names used
      </h1>
      <Field
        adjustFor=""
        className="no-margin-bottom"
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.othernames.branch.help"
        helpActive={false}
        optional={true}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Provide your other names used and the period of time you used them"
        titleSize="h3"
        validate={true}
      >
        <ReactMarkdown
          containerTagName="div"
          parserOptions={Object {}}
          renderers={
            Object {
              "link": [Function],
            }
          }
          source="For example: your maiden name, name(s) by a former marriage, former name(s), alias(es), or nickname(s)."
        />
      </Field>
      <Branch
        adjustFor="buttons"
        className=""
        confirmation="Are you sure you would like to change your answer? This will remove any data currently entered."
        disabled={false}
        label="Have you used any other names?"
        labelSize="h4"
        name="has_othernames"
        noAriaLabel=""
        noLabel="No"
        noValue="No"
        onError={[Function]}
        onUpdate={[Function]}
        optional={false}
        required={false}
        scrollIntoView={false}
        value=""
        warning={true}
        yesAriaLabel=""
        yesLabel="Yes"
        yesValue="Yes"
      />
      <Show
        when={false}
      >
        <Accordion
          appendClass=""
          appendHelp={null}
          appendLabel="Add another name"
          appendMessage={null}
          appendTitle="Do you have additional names to enter?"
          branch={Object {}}
          byline={[Function]}
          caption={null}
          className=""
          closeLabel="Close"
          customDetails={[Function]}
          customSummary={[Function]}
          defaultState={true}
          description="Summary of other names used"
          incomplete="There is a problem with this item, please open it and fix the issue(s)."
          initial={true}
          inject={[Function]}
          items={Array []}
          minimum={1}
          onError={[Function]}
          onUpdate={[Function]}
          openLabel="Open"
          realtime={true}
          removeLabel="Remove"
          required={false}
          scrollToBottom=""
          scrollToTop=""
          skipWarning={false}
          sort={null}
          summary={[Function]}
          timeout={500}
          titleSize="h4"
          transformer={[Function]}
          validator={[Function]}
        >
          <OtherNameItem
            bind={true}
            name="Item"
            onError={[Function]}
            onUpdate={[Function]}
            required={false}
          />
        </Accordion>
      </Show>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <OtherNames
        HasOtherNames={Object {}}
        List={
          Object {
            "branch": Object {},
            "items": Array [],
          }
        }
        defaultState={true}
        dispatch={[Function]}
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        validator={[Function]}
      />,
      "_debugID": 457,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": OtherNames {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleCompletion": [Function],
        "handleError": [Function],
        "props": Object {
          "HasOtherNames": Object {},
          "List": Object {
            "branch": Object {},
            "items": Array [],
          },
          "defaultState": true,
          "dispatch": [Function],
          "onError": [Function],
          "onUpdate": [Function],
          "required": false,
          "validator": [Function],
        },
        "refs": Object {},
        "section": "identification",
        "state": null,
        "store": "Identification",
        "storeKey": "OtherNames",
        "subsection": "othernames",
        "update": [Function],
        "updateBranch": [Function],
        "updateList": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="section-content other-names"
          data-section="identification"
          data-subsection="othernames"
        >
          <h1
            className="section-header"
          >
            Other names used
          </h1>
          <Field
            adjustFor=""
            className="no-margin-bottom"
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.othernames.branch.help"
            helpActive={false}
            optional={true}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your other names used and the period of time you used them"
            titleSize="h3"
            validate={true}
          >
            <ReactMarkdown
              containerTagName="div"
              parserOptions={Object {}}
              renderers={
                Object {
                  "link": [Function],
                }
              }
              source="For example: your maiden name, name(s) by a former marriage, former name(s), alias(es), or nickname(s)."
            />
          </Field>
          <Branch
            adjustFor="buttons"
            className=""
            confirmation="Are you sure you would like to change your answer? This will remove any data currently entered."
            disabled={false}
            label="Have you used any other names?"
            labelSize="h4"
            name="has_othernames"
            noAriaLabel=""
            noLabel="No"
            noValue="No"
            onError={[Function]}
            onUpdate={[Function]}
            optional={false}
            required={false}
            scrollIntoView={false}
            value=""
            warning={true}
            yesAriaLabel=""
            yesLabel="Yes"
            yesValue="Yes"
          />
          <Show
            when={false}
          >
            <Accordion
              appendClass=""
              appendHelp={null}
              appendLabel="Add another name"
              appendMessage={null}
              appendTitle="Do you have additional names to enter?"
              branch={Object {}}
              byline={[Function]}
              caption={null}
              className=""
              closeLabel="Close"
              customDetails={[Function]}
              customSummary={[Function]}
              defaultState={true}
              description="Summary of other names used"
              incomplete="There is a problem with this item, please open it and fix the issue(s)."
              initial={true}
              inject={[Function]}
              items={Array []}
              minimum={1}
              onError={[Function]}
              onUpdate={[Function]}
              openLabel="Open"
              realtime={true}
              removeLabel="Remove"
              required={false}
              scrollToBottom=""
              scrollToTop=""
              skipWarning={false}
              sort={null}
              summary={[Function]}
              timeout={500}
              titleSize="h4"
              transformer={[Function]}
              validator={[Function]}
            >
              <OtherNameItem
                bind={true}
                name="Item"
                onError={[Function]}
                onUpdate={[Function]}
                required={false}
              />
            </Accordion>
          </Show>
        </div>,
        "_debugID": 458,
        "_renderedOutput": <div
          className="section-content other-names"
          data-section="identification"
          data-subsection="othernames"
        >
          <h1
            className="section-header"
          >
            Other names used
          </h1>
          <Field
            adjustFor=""
            className="no-margin-bottom"
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.othernames.branch.help"
            helpActive={false}
            optional={true}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your other names used and the period of time you used them"
            titleSize="h3"
            validate={true}
          >
            <ReactMarkdown
              containerTagName="div"
              parserOptions={Object {}}
              renderers={
                Object {
                  "link": [Function],
                }
              }
              source="For example: your maiden name, name(s) by a former marriage, former name(s), alias(es), or nickname(s)."
            />
          </Field>
          <Branch
            adjustFor="buttons"
            className=""
            confirmation="Are you sure you would like to change your answer? This will remove any data currently entered."
            disabled={false}
            label="Have you used any other names?"
            labelSize="h4"
            name="has_othernames"
            noAriaLabel=""
            noLabel="No"
            noValue="No"
            onError={[Function]}
            onUpdate={[Function]}
            optional={false}
            required={false}
            scrollIntoView={false}
            value=""
            warning={true}
            yesAriaLabel=""
            yesLabel="Yes"
            yesValue="Yes"
          />
          <Show
            when={false}
          >
            <Accordion
              appendClass=""
              appendHelp={null}
              appendLabel="Add another name"
              appendMessage={null}
              appendTitle="Do you have additional names to enter?"
              branch={Object {}}
              byline={[Function]}
              caption={null}
              className=""
              closeLabel="Close"
              customDetails={[Function]}
              customSummary={[Function]}
              defaultState={true}
              description="Summary of other names used"
              incomplete="There is a problem with this item, please open it and fix the issue(s)."
              initial={true}
              inject={[Function]}
              items={Array []}
              minimum={1}
              onError={[Function]}
              onUpdate={[Function]}
              openLabel="Open"
              realtime={true}
              removeLabel="Remove"
              required={false}
              scrollToBottom=""
              scrollToTop=""
              skipWarning={false}
              sort={null}
              summary={[Function]}
              timeout={500}
              titleSize="h4"
              transformer={[Function]}
              validator={[Function]}
            >
              <OtherNameItem
                bind={true}
                name="Item"
                onError={[Function]}
                onUpdate={[Function]}
                required={false}
              />
            </Accordion>
          </Show>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <OtherNames
    HasOtherNames={Object {}}
    List={
      Object {
        "branch": Object {},
        "items": Array [],
      }
    }
    defaultState={true}
    dispatch={[Function]}
    onError={[Function]}
    onUpdate={[Function]}
    required={false}
    validator={[Function]}
  />,
}
`;
